suprb2/optimizer/rule/ns/ns.py

```diff
@@ -261,8 +262,13 @@ def _new_population(self, children: list[Rule], parents: list[Rule], n_rules: in
             return population
 
     def _filter_for_minimal_criteria(self, rules: list[Rule]) -> list[Rule]:
-        if self.MCNS_threshold_matched:
-            rules = [rule for rule in rules if np.count_nonzero(rule.match_) > self.MCNS_threshold_matched]
+        # calculate the 25th percentile value and if it's lower than MNCS_threshold_matched use it instead to filter
+        # a maximum of 25% of the population (to prevent empty populations)
+        maximum_threshold = np.percentile([np.count_nonzero(rule.match_) for rule in rules], 25)
+        if maximum_threshold < self.MCNS_threshold_matched:
+            rules = [rule for rule in rules if np.count_nonzero(rule.match_) >= maximum_threshold]
+        else:
+            rules = [rule for rule in rules if np.count_nonzero(rule.match_) >= self.MCNS_threshold_matched]
```

Comment:
Can be replaced with:
`maximum_threshold = min(np.percentile(self.MCNS_threshold_matched, [np.count_nonzero(rule.match_) for rule in rules], 25))`
`rules = [rule for rule in rules if np.count_nonzero(rule.match_) >= maximum_threshold]`